src/cosalib/vmware.py

```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env python3
+# NOTE: PYTHONUNBUFFERED is set in cmdlib.sh for unbuffered output
+#
+# An operation that mutates a build by generating an ova
+import json
+import os.path
+import shutil
+import sys
+import tarfile
+
+cosa_dir = os.path.dirname(os.path.abspath(__file__))
+sys.path.insert(0, f"{cosa_dir}/cosalib")
+sys.path.insert(0, cosa_dir)
+
+from cosalib.cmdlib import run_verbose
+from cosalib.qemuvariants import QemuVariantImage
+
+
+OVA_TEMPLATE_FILE = '/usr/lib/coreos-assembler/vmware-template.xml'
+
+
+class VmwareImageError(Exception):
+    pass
+
+
+class VmwareOVA(QemuVariantImage):
+    def __init__(self, *args, **kwargs):
+        kwargs["variant"] = "vmware_vmdk"
+        QemuVariantImage.__init__(self, *args, **kwargs)
+
+    def generate_ovf_parameters(self, vmdk, cpu=2,
+                                memory=4096, system_type="vmx-13",
+                                os_type="rhel7_64Guest", scsi="VirtualSCSI",
+                                network="VmxNet3"):
+        """
+        Returns a dictionary with the parameters needed to create an OVF file
+        based on the qemu, vmdk, and info from the build metadata
+        """
+        qemu_info = run_verbose(["qemu-img", "info", vmdk,
+                                 "--output", "json"], capture_output=True)
+        disk_size = json.loads(qemu_info.stdout)['virtual-size']
+        vmdk_size = str(os.stat(vmdk).st_blocks * 512)
+
+        image = self.summary
+        product = f'{self.meta["name"]} {self.summary}'
+        vendor = self.meta['name']
+        version = self.meta['ostree-version']
+
+        params = {
+            'ovf_cpu_count':                    cpu,
+            'ovf_memory_mb':                    memory,
+            'vsphere_image_name':               image,
+            'vsphere_product_name':             product,
+            'vsphere_product_vendor_name':      vendor,
+            'vsphere_product_version':          version,
+            'vsphere_virtual_system_type':      system_type,
+            'vsphere_os_type':                  os_type,
+            'vsphere_scsi_controller_type':     scsi,
+            'vsphere_network_controller_type':  network,
+            'virtual_disk_size':                disk_size,
+            'vmdk_size':                        vmdk_size
+        }
+
+        return params
+
+    def _build_artifacts(self, *args, **kwargs):
+        """
+        Implements the _build_artifact interface to produce an OVA
+        file.
+        """
+        if self.have_artifact and not self.force:
+            raise VmwareImageError(
+                f"{self.image_name} has already been built. {self.force}")
```

Comment:
Hmm, should this end instead with e.g. `Use --force to force a rebuild` ?
Edit: and similarly in the previous commit.

Reply:
Fixed in the latest push for both. I indented that the CLI will handle that part.